README.md
# loopback-swagger

**This module is in Active LTS mode, new features are no longer accepted.**
<br/>(See [Module Long Term Support Policy](#module-long-term-support-policy)
below.)

LoopBack 3 users looking for new features are encouraged to upgrade
to LoopBack 4. Refer to
[loopback-next#1849](https://github.com/strongloop/loopback-next/issues/1849)
for more information on how to upgrade.

## Overview

Utilities to transform between Swagger API specs and LoopBack remoting metadata.

This is an internal module used by the following user-facing tools:

 - [slc loopback:swagger](https://github.com/strongloop/loopback-swagger)
 - [loopback-component-explorer](https://github.com/strongloop/loopback-component-explorer)
 
---
# Swagger-ui v3 with Security

This is a WIP fork of advanced swagger.json OAuth2 support for Loopback using latest swagger-ui (currently 3.9)

## Key Repositories

- [drmikecrowe: loopback-component-explorer](https://github.com/drmikecrowe/loopback-component-explorer)
- [drmikecrowe: loopback-swagger](https://github.com/drmikecrowe/loopback-swagger) 
 
## SwaggerUI v3 Support Status

- [x]  Integrate v3 of the SwaggerUI (using [swagger-ui-dist](https://www.npmjs.com/package/swagger-ui-dist))
- [x]  securityDefinitions Support
- [x]  Integrate with [loopback-component-explorer](https://github.com/strongloop/loopback-component-explorer).
  - [x] Update explorer to use new Swagger-UI
  - [x] Test and document using oauth2-redirect for OAuth2 local development
- [ ] Tested Grant Types
  - [x] implicit
  - [ ] jwt
  - [ ] clientCredentials 
  - [x] authorizationCode 
  - [x] refreshToken 
  - [ ] resourceOwnerPasswordCredentials
- [ ]  Review @raymongfeng realm discussions and determine insure this strategy meets Strongloop plans 
- [ ]  Update automated tests
- [ ]  Review loopback style guide

## Installation

- See [drmikecrowe: loopback-component-explorer README](https://github.com/drmikecrowe/loopback-component-explorer) for current installation instructions

### Development redirectURIs

For development only, you may add the following redirectURI to your applications object (assuming your API is running on localhost:3001, and your authentication server is running on localhost:3000):

    "redirectURIs" : [
        "http://localhost:3001/explorer/oauth2-redirect.html"
    ], 

### Add detailed API info to your `component-config`

In your component-config.local.js, OAuth2 configuration may be specified as follows:

```js
var path = require('path');

module.exports = {
  "loopback-component-explorer":        {
    "mountPath":           "/explorer",
    "uiDirs":              "explorer",
    "apiInfo":             {
      "title":          "Loopback API Explorer",
      "description":    "A description of my API",
      "version":        require('../package.json').version,
      "termsOfService": "http://my-super-site.com/terms",
      "contact":        {
        "name":  "Support Team",
        "email": "support@my-super-site.com"
      }
    },
    "host":                "my-super-site.com",
    "schemes":             [
      "http",
      "https"
    ],
    "consumes":            [
      "application/json"
    ],
    "produces":            [
      "application/json",
      "application/xml"
    ],
    "securityDefinitions": {
      "OauthPassword": {
        "type":             "oauth2",
        "flow":             "implicit",
        "description":      "OAuth2 Password authentication",
        "authorizationUrl": path.join("/oauth/authorize"),
        "scopes":           {
          "read": "Access API methods",
          "write": "Update API methods"
        }
      },
      "OauthAccessCode": {
        "type":             "oauth2",
        "description":      "OAuth2 Access Code authentication",
        "flow":             "accessCode",
        "authorizationUrl": path.join("/oauth/authorize"),
        "tokenUrl":         path.join("/oauth/token"),
        "scopes":           {
          "read": "Access API methods",
          "write": "Update API methods"
        }
      }
    },
    "security":            [
      {
        "OauthAccessCode": ["read", "write"],
        "OauthPassword": ["read", "write"]
      }
    ]
  }
};
```

I implemented via component-config.local.js in order to allow the _path_ module in case the authorization server were needed to be a separate URL.

This produces swagger.yaml such as the following:

```yaml
info:
  title: Loopback API Explorer
  description: A description of my API
  version: 0.1.26
  contact:
    email: support@my-super-site.com
    name: Support Team
  termsOfService: 'http://my-super-site.com/terms'
basePath: /api/v1
host: my-super-site.com
consumes:
  - application/json
produces:
  - application/json
  - application/xml
securityDefinitions:
  OauthPassword:
    type: oauth2
    flow: implicit
    description: OAuth2 Password authentication
    authorizationUrl: /oauth/authorize
    scopes:
      read: Access API methods
      write: Update API methods
  OauthAccessCode:
    type: oauth2
    description: OAuth2 Access Code authentication
    flow: accessCode
    authorizationUrl: /oauth/authorize
    tokenUrl: /oauth/token
    scopes:
      read: Access API methods
      write: Update API methods
security:
  - OauthAccessCode:
      - read
      - write
    OauthPassword:
      - read
      - write
paths: {}
definitions: {}
tags: {}
```

### Route Specific Scopes

*IF* you need to specify a specific OAuth2 Scope at a Loopback method level, this project supports that functionality in a limited fashion:

* The swagger.json will syntatically be correct per the JSON schema definition for Swagger v2.
* It does **NOT** implement a custom role-resolver for that specific method.  That will be left to each application.

To specify a specific scope on a single route, in the model definition file:

```json
{
  "acls": [
    {
      "accessType": "EXECUTE",
      "principalType": "OauthPassword",
      "principalId": "createSomething",
      "scope": "write",
      "permission": "ALLOW"
    }
  ]
}
```

will produce Swagger yaml such as:

```yaml
  '/MyModel/createSomething':
    post:
      tags:
        - MyModel
      operationId: MyModel.createSomething
      parameters:
        - name: data
          in: body
          description: >-
            Data object to be saved
          required: true
          schema:
            description: >-
              Data object to be saved
            $ref: '#/definitions/SomethingObject'
      responses:
        '200':
          description: Request was successful
          schema:
            $ref: '#/definitions/Something'
      deprecated: false
      security:
        - OauthPassword:
            - write
```

 - [slc loopback:swagger](https://docs.strongloop.com/display/LB/Swagger+generator)
 - [loopback-explorer](https://github.com/strongloop/loopback-explorer)

## Module Long Term Support Policy

This module adopts the [
Module Long Term Support (LTS)](http://github.com/CloudNativeJS/ModuleLTS) policy,
with the following End Of Life (EOL) dates:

| Version | Status          | Published | EOL      |
| ------- | --------------- | --------- | -------- |
| 5.x     | Active LTS      | Sep 2017  | Dec 2019 |
| 4.x     | Maintenance LTS | Feb 2017  | Apr 2019 |
| 3.x     | Maintenance LTS | Dec 2016  | Apr 2019 |

Learn more about our LTS plan in [docs](https://loopback.io/doc/en/contrib/Long-term-support.html).


CHANGES.md
2019-01-07, Version 5.7.3

 * Merge securityDefinitions back into latest

2019-01-07, Version 5.7.2
=========================

 * Fix route helper to ignore status/header retvals (Y.Shing)


2018-10-18, Version 5.7.1
=========================

 * README: update LTS status (Miroslav Bajtoš)


2018-08-30, Version 5.7.0
=========================

 * specgen: Retrieve description from package.json (Melle Boersma)

 * Upgrade eslint-config-loopback + fix formatting (Miroslav Bajtoš)


2018-07-10, Version 5.6.0
=========================

 * specgen: emit correct non-root return types (Dan Jarvis)

 * Update mocha & chai to latest (Miroslav Bajtoš)

 * Disable package-lock feature of npm (Miroslav Bajtoš)

 * Update eslint + config to latest (Miroslav Bajtoš)

 * Update strong-globalize and debug to latest (Miroslav Bajtoš)

 * Travis: add Node.js 8.x and 10.x (Miroslav Bajtoš)

 * Drop support for Node.js 4.x (Miroslav Bajtoš)

 * [WebFM] cs/pl/ru translation (candytangnb)


2017-12-04, Version 5.5.0
=========================

 * Improve support for $ref (Raymond Feng)


2017-11-21, Version 5.4.0
=========================

 * Handle embedded schemas for parameter/response (Raymond Feng)


2017-11-21, Version 5.3.0
=========================

 * Add support for examples in models and responses (Zak Barbuto)


2017-11-20, Version 5.2.2
=========================

 * Improve swagger mapping (Raymond Feng)

 * chore: update license (Diana Lau)


2017-10-27, Version 5.2.1
=========================

 * Fix file upload attachments (Zak Barbuto)


2017-10-27, Version 5.2.0
=========================

 * feat(route-helper): Add 'documented' flag for hiding params (Samuel Reed)


2017-10-13, Version 5.1.0
=========================

 * update strong-globalize to 3.1.0 (shimks)

 * CODEOWNERS: add zbarbuto (Miroslav Bajtoš)

 * update globalize string (Diana Lau)

 * fix basePath, add store.json (wing328)


2017-09-05, Version 5.0.0
=========================

 * Enable 'updateOnly' feature based on generateOperationScopedModels (#99) (Rashmi Hunt)

 * Bump version to 5.0.0 (Raymond Feng)

 * Set required to be true for path parameters (Raymond Feng)


2017-08-22, Version 4.2.0
=========================

 * Support updateOnly feature (#92) (Rashmi Hunt)

 * Add stalebot configuration (Kevin Delisle)

 * Update Issue and PR Templates (#95) (Sakib Hasan)


2017-08-14, Version 4.1.0
=========================

 * Allow externalDocs and custom names for models (Zak Barbuto)

 * Add support for file-type parameters (Zak Barbuto)

 * Add community maintainers to CODEOWNERS (Miroslav Bajtoš)

 * Add CODEOWNER file (Diana Lau)

 * Update eslint config to the latest (Miroslav Bajtoš)


2017-04-27, Version 4.0.3
=========================

 * type-registry: add DateString type (Kevin Delisle)

 * Travis CI config (Kevin Delisle)


2017-03-02, Version 4.0.2
=========================

 * Allow swagger 1.2 (Raymond Feng)


2017-03-02, Version 4.0.1
=========================

 * Use a default model is added if no tags found (Raymond Feng)

 * Replicate new issue_template from loopback (Siddhi Pai)

 * Replicate issue_template from loopback repo (Siddhi Pai)


2017-02-01, Version 4.0.0
=========================

 * Bump version to 4.0.0 (Raymond Feng)

 * Remove .npmignore (Raymond Feng)

 * Enhance Swagger to LoopBack mapping for code gen (Raymond Feng)

 * Upgrade eslint-config to 7.x, eslint to 3.x (Miroslav Bajtoš)


2016-12-22, Version 3.0.1
=========================

 * Omit null default values from schema (Heath Morrison)


2016-12-21, Version 3.0.0
=========================

 * fix: package.json to reduce vulnerabilities (Ryan Graham)

 * Update paid support URL (Siddhi Pai)

 * Adjust route parameters sequence (Tao Yuan)

 * Start 3.x + drop support for Node v0.10/v0.12 (siddhipai)

 * Drop support for Node v0.10 and v0.12 (Siddhi Pai)

 * Start the development of the next major version (Siddhi Pai)

 * Coerce form to formData in accepts.http.source (Simon Ho)

 * Fix date output format to "date-time" (Stefan B)


2016-10-13, Version 2.8.0
=========================

 * Update pt translation file (Candy)

 * Update ja translation file (Candy)

 * Update translation files - round#2 (Candy)

 * Add translated files (gunjpan)

 * Update deps to loopback 3.0.0 RC (Miroslav Bajtoš)

 * Remove juggler as a dependency (Amir Jafarian)


2016-09-13, Version 2.7.0
=========================

 * Allow setting null on error's responseModel (Alex Plescan)

 * Use juggler@3 and loopback@3 for testing (Amir Jafarian)


2016-09-07, Version 2.6.0
=========================

 * Allow object arguments to provide Model (Miroslav Bajtoš)

 * Expose object types as nested properties (Carl Fürstenberg)


2016-09-05, Version 2.5.0
=========================

 * Add `patch` verb in test (Amir Jafarian)

 * Add globalization. (Richard Pringle)

 * Update URLs in CONTRIBUTING.md (#49) (Ryan Graham)

 * Add doc root (jannyHou)

 * Add custom definition to swagger file (jannyHou)


2016-05-06, Version 2.4.3
=========================

 * update copyright notices and license (Ryan Graham)


2016-04-13, Version 2.4.2
=========================

 * Do not generate required if no item is present (Raymond Feng)

 * Fix linting errors (Amir Jafarian)

 * Auto-update by eslint --fix (Amir Jafarian)

 * Add eslint infrastructure (Amir Jafarian)


2016-04-07, Version 2.4.1
=========================

 * Disable the warning as it prints messages during loopback scaffolding (Raymond Feng)


2016-04-07, Version 2.4.0
=========================

 * improve scheme generation for return object (hideya kawahara)

 * Support default status codes (Candy)


2016-03-19, Version 2.3.2
=========================

 * Handle {id} and . in operation id (Raymond Feng)


2016-03-03, Version 2.3.1
=========================

 * Handle extensions under paths (Raymond Feng)


2016-02-09, Version 2.3.0
=========================

 * Make type geopoint case insensitive (Candy)


lib/specgen/model-helper.js
// Copyright IBM Corp. 2015,2016. All Rights Reserved.
// Node module: loopback-swagger
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

'use strict';

/**
 * Module dependencies.
 */
var schemaBuilder = require('./schema-builder');
var typeConverter = require('./type-converter');
var TypeRegistry = require('./type-registry');
var _ = require('lodash');

/**
 * Export the modelHelper singleton.
 */
var modelHelper = module.exports = {
  /**
   * Given a class (from remotes.classes()), generate a model definition.
   * This is used to generate the schema at the top of many endpoints.
   * @param  {Class} modelClass Model class.
   * @param {TypeRegistry} typeRegistry Registry of types and models.
   * @return {Object} Associated model definition.
   */
  registerModelDefinition: function(modelCtor, typeRegistry, securityDefinitions, opts) {
    var lbdef = modelCtor.definition;

    if (!lbdef) {
      // The model does not have any definition, it was most likely
      // created as a placeholder for an unknown property type
      return;
    }

    var name = lbdef.name;
    if (typeRegistry.isDefined(name)) {
      // The model is already included
      return;
    }

    typeRegistry.registerModel(name, function() {
      return definitionFunction(modelCtor, typeRegistry, securityDefinitions);
    });

    // check if ModelClass has defined getUpdateOnlyProperties() function to
    // avoid version issues
    if (opts && opts.generateOperationScopedModels && modelCtor.getUpdateOnlyProperties) {
      const excludeProps = modelCtor.getUpdateOnlyProperties();
      // if at least one excludeProp is found, we need to create another model
      // to be used only for create operation and this model will not have
      // excludeProp included in the model. e.g generated "id" property
      if (excludeProps && excludeProps.length > 0) {
        const modelName = '$new_' + name;
        typeRegistry.registerModel(modelName, function() {
          return definitionFunction(modelCtor, typeRegistry, {excludeProps});
        });
      }
    }
  },

  isHiddenProperty: function(definition, propName) {
    return definition.settings &&
      Array.isArray(definition.settings.hidden) &&
      definition.settings.hidden.indexOf(propName) !== -1;
  },
};

var definitionFunction = function(modelCtor, typeRegistry, securityDefinitions, options) {
  var lbdef = modelCtor.definition;

  var swaggerDef = {
    description: typeConverter.convertText(
      lbdef.description || (lbdef.settings && lbdef.settings.description)
    ),
    properties: {},
    required: [],
  };

  if (lbdef.settings && lbdef.settings.swagger && lbdef.settings.swagger.example) {
    swaggerDef.example = lbdef.settings.swagger.example;
  }

  addSwaggerExtensions(lbdef.settings);
  addSwaggerSecurity(lbdef, typeRegistry, securityDefinitions);

  var properties = lbdef.rawProperties || lbdef.properties;

  // Iterate through each property in the model definition.
  // Types may be defined as constructors (e.g. String, Date, etc.),
  // or as strings; swaggerSchema.buildFromLoopBackType() will take
  // care of the conversion.
  Object.keys(properties).forEach(function(key) {
    var prop = properties[key];

    // Hide hidden properties.
    if (modelHelper.isHiddenProperty(lbdef, key))
      return;

    // Get a type out of the constructors we were passed.
    var schema = schemaBuilder.buildFromLoopBackType(prop, typeRegistry);

    var desc = typeConverter.convertText(prop.description || prop.doc);
    if (desc) schema.description = desc;

    // Required props sit in a per-model array.
    if (prop.required || (prop.id && !prop.generated)) {
      swaggerDef.required.push(key);
    }

    // if model has excludeProps properties, skip adding to the model since this
    // model is used for create operation only where this property
    // should not exist
    if (options && options.excludeProps && _.includes(options.excludeProps, key)) {
      return;
    }
    swaggerDef.properties[key] = schema;
  });

  if (lbdef.settings) {
    var strict = lbdef.settings.strict;
    var additionalProperties = lbdef.settings.additionalProperties;
    var notAllowAdditionalProperties = strict || (additionalProperties !== true);
    if (notAllowAdditionalProperties) {
      swaggerDef.additionalProperties = false;
    }
  }

  if (!swaggerDef.required.length) {
    // "required" must have at least one item when present
    delete swaggerDef.required;
  }

  // Add models from settings
  if (lbdef.settings && lbdef.settings.models) {
    for (var m in lbdef.settings.models) {
      var model = modelCtor[m];
      if (typeof model !== 'function' || !model.modelName) continue;
      modelHelper.registerModelDefinition(model, typeRegistry);
      typeRegistry.reference(model.modelName);
    }
  }

  // Generate model definitions for related models
  /* eslint-disable one-var */
  for (var r in modelCtor.relations) {
    var rel = modelCtor.relations[r];
    if (rel.modelTo) {
      modelHelper.registerModelDefinition(rel.modelTo, typeRegistry);
    }
    if (rel.modelThrough) {
      modelHelper.registerModelDefinition(rel.modelThrough, typeRegistry);
    }
  }
  /* eslint-enable one-var */
  return swaggerDef;

  function addSwaggerExtensions(defs) {
    for (var def in defs) {
      if (def.match(/^x\-/)) {
        swaggerDef[def] = defs[def];
      }
    }
  };

  function addSwaggerSecurity(lbdef, typeRegistry, securityDefinitions) {
    var modelName = lbdef.name;
    var acls = lbdef.settings && lbdef.settings.acls ? lbdef.settings.acls : {};
    for (var i in acls) {
      var acl = acls[i];
      if (acl.principalType in securityDefinitions) {
        typeRegistry.routeSecurity(modelName, acl);
      }
    }
  };
};



lib/specgen/route-helper.js
// Copyright IBM Corp. 2015,2016. All Rights Reserved.
// Node module: loopback-swagger
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

'use strict';

// Globalization
var g = require('strong-globalize')();

/**
 * Module dependencies.
 */

var debug = require('debug')('loopback:explorer:routeHelpers');
var _assign = require('lodash').assign;
var typeConverter = require('./type-converter');
var schemaBuilder = require('./schema-builder');

/**
 * Export the routeHelper singleton.
 */
var routeHelper = module.exports = {
  /**
   * Given a route, generate an API description and add it to the doc.
   * If a route shares a path with another route (same path, different verb),
   * add it as a new operation under that path entry.
   *
   * Routes can be translated to API declaration 'operations',
   * but they need a little massaging first. The `accepts` and
   * `returns` declarations need some basic conversions to be compatible.
   *
   * This method will convert the route and add it to the doc.
   *
   * @param  {Route} route    Strong Remoting Route object.
   * @param  {Class} classDef Strong Remoting class.
   * @param  {TypeRegistry} typeRegistry Registry of types and models.
   * @param  {Object} operationIdRegistry Registry of operationIds mapping
   *  operationId to an operation object.
   * @param  {Object} paths   Swagger Path Object,
   *   see https://github.com/swagger-api/swagger-spec/blob/master/versions/2.0.md#pathsObject
   */
  addRouteToSwaggerPaths: function(route, classDef, typeRegistry,
    operationIdRegistry, paths, opts) {
    var entryToAdd = routeHelper.routeToPathEntry(route, classDef,
      typeRegistry,
      operationIdRegistry, opts);
    if (!(entryToAdd.path in paths)) {
      paths[entryToAdd.path] = {};
    }
    paths[entryToAdd.path][entryToAdd.method] = entryToAdd.operation;
  },

  /**
   * Massage route.accepts.
   * @param  {Object} route    Strong Remoting Route object.
   * @param  {Class}  classDef Strong Remoting class.
   * @param  {TypeRegistry} typeRegistry Registry of types and models.
   * @return {Array}           Array of param docs.
   */
  convertAcceptsToSwagger: function(route, classDef, typeRegistry, opts) {
    var accepts = route.accepts || [];
    var split = route.method.split('.');
    if (classDef && classDef.sharedCtor &&
        classDef.sharedCtor.accepts && split.length > 2 /* HACK */) {
      accepts = [].concat(classDef.sharedCtor.accepts)
        .concat(accepts);
    }

    // Filter out parameters that are generated from the incoming request,
    // or generated by functions that use those resources.
    accepts = accepts.filter(function(arg) {
      // Allow undocumenting a param.
      if (arg.documented === false) return false;
      // Below conditions are only for 'http'
      if (!arg.http) return true;
      // Don't show derived arguments.
      if (typeof arg.http === 'function') return false;
      // Don't show arguments set to the incoming http request.
      // Please note that body needs to be shown, such as User.create().
      if (arg.http.source === 'req' ||
        arg.http.source === 'res' ||
        arg.http.source === 'context') {
        return false;
      }
      return true;
    });

    // Turn accept definitions in to parameter docs.
    accepts = accepts.map(
      routeHelper.acceptToParameter(route, classDef, typeRegistry, opts)
    );

    return accepts;
  },

  /**
   * Massage route.returns.
   * @param  {Object} route    Strong Remoting Route object.
   * @return {Object}          A single returns param doc.
   */
  convertReturnsToSwagger: function(route, typeRegistry) {
    var routeReturns = route.returns;
    if (!routeReturns || !routeReturns.length) {
      // An operation that returns nothing will have
      // no schema declaration for its response.
      return undefined;
    }

    // Filter out arguments having http.target set to 'header' or 'status'
    routeReturns = routeReturns.filter(function(arg) {
      const target = arg.http && arg.http.target;
      return target !== 'header' && target !== 'status';
    });

    if (routeReturns.length === 1 && routeReturns[0].root) {
      if (routeReturns[0].model) {
        return {$ref: typeRegistry.reference(routeReturns[0].model)};
      }
      return schemaBuilder.buildFromLoopBackType(routeReturns[0], typeRegistry);
    } else if (routeReturns.length === 1 && routeReturns[0].type === 'ReadableStream') {
      return {type: 'file'};
    }

    // Construct scheme for the return object
    var schema = {type: 'object'};
    schema.properties = {};
    routeReturns.forEach(function(ret) {
      var propName = ret.name || ret.arg;
      var idlType = schemaBuilder.getLdlTypeName(ret.type);
      // Take care of array which can be nested.
      // Note that we cannot simply use buildFromLoopBackType since it converts unknown type to
      // '$ref': '#/definitions/UnknownType', whereas we decided to emit 'type: object' for such a case.
      // See https://github.com/strongloop/loopback-swagger/pull/28#discussion_r54873911
      // The following code is needed to take care of a nested array of an unknown type.
      var itemIdlType = idlType;
      var genericIdlType = {type: 'object'};
      while (Array.isArray(itemIdlType)) {
        itemIdlType = schemaBuilder.getLdlTypeName(itemIdlType[0]);
        genericIdlType = {type: 'array', items: genericIdlType};
      }
      if (schemaBuilder.isPrimitiveType(itemIdlType)) {
        schema.properties[propName] =
          schemaBuilder.buildFromLoopBackType(ret, typeRegistry);
      } else {
        if (typeRegistry.isDefined(itemIdlType)) {
          schema.properties[propName] =
            schemaBuilder.buildFromLoopBackType(ret, typeRegistry);
        } else {
          debug('Swagger: temporarily using `object` instead of unknown ' +
          'type %j found in route %j', itemIdlType, route);
          schema.properties[propName] = genericIdlType;
        }
      }
      if (ret.required) {
        if (schema.required == null) {
          schema.required = [];
        }
        schema.required.push(propName);
      }
    });

    return schema;
  },

  /**
   * Converts from an sl-remoting-formatted "Route" description to a
   * Swagger-formatted "Path Item Object"
   * See swagger-spec/2.0.md#pathItemObject
   */
  routeToPathEntry: function(route, classDef,
    typeRegistry, operationIdRegistry, opts) {
    // Some parameters need to be altered; eventually most of this should
    // be removed.
    var accepts = routeHelper.convertAcceptsToSwagger(route, classDef,
      typeRegistry, opts);
    var consumes = undefined;
    for (var i = 0; i < accepts.length; i++) {
      if (accepts[i].in === 'formData') {
        consumes = ['application/x-www-form-urlencoded'];
        break;
      }
    }
    var returns = routeHelper.convertReturnsToSwagger(route, typeRegistry);
    var statusCode = route.returns && route.returns.length ? 200 : 204;

    if (route.http && route.http.status) {
      statusCode = route.http.status;
    }

    var responseMessages = {};
    responseMessages[statusCode] = {
      description: 'Request was successful',
      schema: returns,
      // TODO - headers, examples
    };

    if (route.returns && route.returns[0] && route.returns[0].example) {
      responseMessages[statusCode].examples = responseMessages[statusCode].examples || {};
      responseMessages[statusCode].examples['application/json'] = route.returns[0].example;
    }

    if (route.errors) {
      // TODO define new LDL syntax that is status-code-indexed
      // and which allow users to specify headers & examples
      route.errors.forEach(function(msg) {
        var schema = null;
        if (msg.responseModel) {
          schema = schemaBuilder.buildFromLoopBackType(msg.responseModel,
            typeRegistry);
        }
        responseMessages[msg.code] = {
          description: msg.message,
          schema: schema,
          // TODO - headers, examples
        };
      });
    }

    if (route.http) {
      var errorStatus = route.http.errorStatus;
      if (!responseMessages[errorStatus]) {
        responseMessages[errorStatus] = {
          description: 'Unknown error',
          // TODO - headers, examples
        };
      }
    }

    debug('route %j', route);

    var path = routeHelper.convertPathFragments(route.path);
    var verb = routeHelper.convertVerb(route.verb);

    var tags = [];
    var swaggerSettings = classDef && classDef.ctor && classDef.ctor.settings &&
      classDef.ctor.settings.swagger || {};

    if (swaggerSettings.tag && swaggerSettings.tag.name) {
      tags.push(swaggerSettings.tag.name);
    } else if (classDef && classDef.name) {
      tags.push(classDef.name);
    }

    var operationId = createUniqueOperationId(route.method, verb, path,
      operationIdRegistry);
    var entry = {
      path: path,
      method: verb,
      operation: {
        tags: tags,
        summary: typeConverter.convertText(route.description),
        description: typeConverter.convertText(route.notes),
        operationId: operationId,
        // [drmikecrowe] we need a special consumes for formData with swagger-v3
        consumes: consumes,
        // [bajtos] we are omitting produces, as they are same
        // for all methods and they are already specified in top-level fields
        parameters: accepts,
        responses: responseMessages,
        deprecated: !!route.deprecated,
        // TODO: security
      },
    };

    operationIdRegistry[operationId] = entry;

    return entry;
  },

  convertPathFragments: function convertPathFragments(path) {
    return path.split('/').map(function(fragment) {
      if (fragment.charAt(0) === ':') {
        return '{' + fragment.slice(1) + '}';
      }
      return fragment;
    }).join('/');
  },

  convertVerb: function convertVerb(verb) {
    if (verb.toLowerCase() === 'all') {
      return 'post';
    }

    if (verb.toLowerCase() === 'del') {
      return 'delete';
    }

    return verb.toLowerCase();
  },

  /**
   * A generator to convert from an sl-remoting-formatted "Accepts" description
   * to a Swagger-formatted "Parameter" description.
   */
  acceptToParameter: function acceptToParameter(route, classDef, typeRegistry, opts) {
    var DEFAULT_TYPE =
      route.verb.toLowerCase() === 'get' ? 'query' : 'formData';

    return function(accepts) {
      var name = accepts.name || accepts.arg;
      var paramType = DEFAULT_TYPE;

      // TODO: Regex. This is leaky.
      if (route.path.indexOf(':' + name) !== -1) {
        paramType = 'path';
      }

      // Check the http settings for the argument
      if (accepts.http && accepts.http.source) {
        paramType = accepts.http.source === 'form' ?
          'formData' :
          accepts.http.source;
      }

      // TODO: ensure that paramType has a valid value
      //  path, query, header, body, formData
      // See swagger-spec/2.0.md#parameterObject

      var paramObject = {
        name: name,
        in: paramType,
        description: typeConverter.convertText(accepts.description),
        // For path parameters, required must be true
        required: paramType === 'path' ? true : !!accepts.required,
      };

      var schema = schemaBuilder.buildFromLoopBackType(accepts, typeRegistry, opts);
      if (paramType === 'body') {
        // HACK: Derive the type from model
        if (paramObject.name === 'data' && schema.type === 'object') {
          paramObject.schema = {$ref: typeRegistry.reference(classDef.name)};
        } else {
          paramObject.schema = schema;
        }
      } else if (schema.type === 'file') {
        paramObject.type = 'file';
        paramObject.in = 'formData';
        paramObject.allowMultiple = false;
        paramObject.description = 'File to upload';
      } else {
        var isComplexType = schema.type === 'object' ||
                            schema.type === 'array' ||
                            schema.$ref;
        if (isComplexType) {
          paramObject.type = 'string';
          paramObject.format = 'JSON';
          // TODO support array of primitive types
          // and map them to Swagger array of primitive types
        } else {
          _assign(paramObject, schema);
        }
      }

      return paramObject;
    };
  },
};

function createUniqueOperationId(methodName, verb, path, operationIdRegistry) {
  // [bajtos] We used to remove leading model name from the operation
  // name for Swagger Spec 1.2. Swagger Spec 2.0 requires
  // operation ids to be unique, thus we have to include the model name.
  var id = methodName;

  if (!(id in operationIdRegistry)) {
    // The id is already unique
    return id;
  }

  var baseId = id;
  id = createLongOperationId(baseId, verb, path);
  if (id in operationIdRegistry) {
    g.warn('Warning: detected multiple remote methods ' +
      'at the same HTTP endpoint. ' +
      '{{Swagger operation ids}} will NOT be unique.');
  }

  // Rename the first operation so that all operation ids of
  // a multi-endpoint method are consistently using the long form
  if (operationIdRegistry[baseId]) {
    var oldEntry = operationIdRegistry[baseId];
    var newId = createLongOperationId(baseId, oldEntry.method, oldEntry.path);
    oldEntry.operation.operationId = newId;
    operationIdRegistry[newId] = oldEntry;
    operationIdRegistry[baseId] = null;
  }

  return id;
}

function createLongOperationId(baseId, verb, path) {
  return baseId + '__' + verb + path.replace(/[\/:]+/g, '_');
}
